File: src/commands/advise.ts
import { Flags } from "@oclif/core";

import { QuoinCommand } from "../base.js";
import { advise } from "../core/auditor.js";
import type { Advice } from "../core/auditor.js";
import { auditInputs } from "../core/evidence.js";
import { loadMethodCatalog } from "../method-catalog.js";
import { coverage, propertyShapes } from "../core/quire.js";

export default class Advise extends QuoinCommand {
  static summary =
    "Recommend a verification method for each obligation, from the catalog.";
  static description = `The deterministic half of the test-plan advisor (FR-031). Rules match or they do
not; where they are inconclusive it says so and stops, rather than guessing.

Before this, the method table was skill-local prose and \`Verification\` columns
defaulted to Test by habit — nothing ever advised DAST for an attack surface,
monitors for a temporal property, or fault injection for a reliability NFR.

Obligations come from quire (coverage --json), and each criterion's FR-052
property shape from quire (properties --json), so the advice is over what the
spec actually states rather than over a document somebody pasted in.

Read the output as a recommendation, never a verdict. An LLM may judge the
residue afterwards — labelled as judgement (the FR-042 / ADR-0010 discipline).`;

  static examples = [
    "quoin advise",
    "quoin advise --mismatch-only",
    "quoin advise --json",
  ];

  static flags = {
    repo: Flags.string({ description: "Repository root.", default: "." }),
    module: Flags.string({
      description: "Module directory supplying the traceability model.",
    }),
    "mismatch-only": Flags.boolean({
      description:
        "Only genuine disagreements: the authored method is a declared " +
        "method or class and is not among the recommendations. Uncatalogued " +
        "values are not mismatches. Combines with --inconclusive-only as a " +
        "union.",
    }),
    "inconclusive-only": Flags.boolean({
      description:
        "Only obligations no rule matched. Combines with --mismatch-only as " +
        "a union.",
    }),
    json: Flags.boolean({ description: "Emit the advice as JSON." }),
  };

  async run(): Promise<void> {
    const { flags } = await this.parse(Advise);

    const derived = coverage(
      flags.repo,
      flags.module ? [flags.module] : undefined,
    );

    const catalog = loadMethodCatalog(
      flags.module ? [flags.module] : undefined,
    );
    if (catalog.methods.length === 0) {
      this.error(
        "no active module declares a `verification_catalog`, so there is " +
          "nothing to advise from. Install one (e.g. spec-artifacts-process) " +
          "or pass --module.",
        { exit: 2 },
      );
    }
    for (const bad of catalog.unreadable) {
      this.warn(`skipped ${bad.moduleRoot}: ${bad.reason}`);
    }

    const { shapes, unresolved } = propertyShapes(flags.repo);
    for (const document of unresolved) {
      // Named, not counted. A document that resolved to no archetype is a
      // criterion nothing classified, and the advice over it rests on
      // statement text alone.
      this.warn(`no archetype resolved for ${document}; not classified.`);
    }
    const obligations = derived.obligations;
    if (obligations.length > 0 && Object.keys(shapes).length === 0) {
      // Said out loud rather than degraded silently. `property_shapes` is the
      // axis most catalog entries are keyed on, so without it `round-trip`
      // never reaches property-based or metamorphic testing and the advice is
      // characteristics-only — which looks like "no rule matched" and is not.
      this.warn(
        "no property shapes were read, so recommendations rest on statement " +
          "text alone. `quire properties` resolves archetypes from the whole " +
          "installed module set; check that the modules declaring FR/NFR/StR " +
          "are installed (quoin module list).",
      );
    }

    // The store is read ONCE, here, and handed in. The advisor performs no
    // I/O — an advisor that could reach the filesystem could also disagree
    // with the auditor about what it found (ADR-0011).
    const store = auditInputs(flags.repo);

    // ONE request for the whole population. The retained command built each
    // obligation's facts itself and called `advise` per obligation, which is
    // fine inside one process and wrong across a boundary: the unit of IPC is
    // a command-shaped operation, so this is one round trip and not N.
    //
    // The uncatalogued-method join (quoin#168) went with it: quire's own
    // diagnosis of which authored values the catalog never declared, keyed by
    // the `value` field byte-equal to `Obligation.method` (quire-rs CR-091).
    const advised = advise({
      catalog,
      obligations,
      shapes,
      bindings: store.bindings,
      runs: store.runs,
      diagnostics: derived.diagnostics,
    });
    const advice = advised.advice;
    if (advised.degraded) {
      this.warn(
        "engine predates vocabulary classification: this quire's " +
          "`uncatalogued-verification-method` diagnostics carry no `value`, " +
          "so an authored method the catalog never declared cannot be told " +
          "from a genuine disagreement. Every disagreement is reported as a " +
          "mismatch, as before. Update quire-cli to restore the three-state " +
          "split (quire-rs CR-091).",
      );
    }

    // `--*-only` flags combine as a UNION. Intersection made the pair
    // `--mismatch-only --inconclusive-only` a guaranteed zero rows (a mismatch
    // requires recommendations, inconclusive means none), which read as "all
    // clear" over a corpus full of findings.
    const only: Array<(a: Advice) => boolean> = [];
    if (flags["mismatch-only"]) only.push((a) => a.mismatch);
    if (flags["inconclusive-only"]) only.push((a) => a.inconclusive);
    const shown =
      only.length === 0
        ? advice
        : advice.filter((a) => only.some((wanted) => wanted(a)));

    if (flags.json) {
      this.log(JSON.stringify({ advice: shown }, null, 2));
      return;
    }
    this.render(shown, advice);
  }

  private render(shown: Advice[], all: Advice[]): void {
    for (const a of shown) {
      const head = a.recommended
        .slice(0, 3)
        .map((r) => `${r.method} (${r.reasons.map((x) => x.value).join(", ")})`)
        .join("; ");
      this.log(
        `${a.obligation}  authored=${a.authored ?? "—"}  ` +
          (a.inconclusive ? "inconclusive" : `recommend: ${head}`) +
          (a.mismatch ? "  ⚠ mismatch" : "") +
          (a.uncatalogued ? "  ⚠ uncatalogued" : ""),
      );
    }
    this.log("");
    // Tallies count the FULL population, whatever was filtered out: a footer
    // that tallied only shown rows understated the very totals the filter was
    // asked about (`0 of 33 shown — 0 mismatch, 0 inconclusive`).
    const count = (state: (a: Advice) => boolean) => all.filter(state).length;
    this.log(
      `${shown.length} of ${all.length} obligation(s) shown. ` +
        `Of all ${all.length}: ${count((a) => a.mismatch)} mismatch, ` +
        `${count((a) => a.uncatalogued)} uncatalogued, ` +
        `${count((a) => a.inconclusive)} inconclusive. ` +
        `Recommendations, not verdicts: confirm the method in spec review.`,
    );
  }
}

File: rust/crates/quoin-cli/src/advise.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Agent-IX

//! Rust adapter for retained `quoin advise`.

use clap::{Arg, ArgAction, ArgMatches, Command};
use quoin_core::protocol::Response;

use crate::core_bridge::invoke;

pub(crate) fn command() -> Command {
    Command::new("advise")
        .about("Recommend a verification method for each obligation, from the catalog")
        .arg(Arg::new("repo").long("repo").default_value("."))
        .arg(Arg::new("module").long("module"))
        .arg(
            Arg::new("mismatch-only")
                .long("mismatch-only")
                .action(ArgAction::SetTrue),
        )
        .arg(
            Arg::new("inconclusive-only")
                .long("inconclusive-only")
                .action(ArgAction::SetTrue),
        )
        .arg(Arg::new("json").long("json").action(ArgAction::SetTrue))
}

pub(crate) fn run(arguments: &ArgMatches) -> Result<Response, String> {
    let repo = required(arguments, "repo")?;
    let modules = arguments
        .get_one::<String>("module")
        .map_or_else(Vec::new, |module| vec![module.clone()]);
    let coverage = invoke(
        "quire.coverage",
        &serde_json::json!({ "scope": repo, "modules": modules }),
    )?;
    if !coverage.outcome.carries_payload() {
        return Ok(coverage);
    }
    let properties = invoke(
        "quire.properties",
        &serde_json::json!({ "scope": repo, "modules": [], "documents": [] }),
    )?;
    if !properties.outcome.carries_payload() {
        return Ok(properties);
    }
    let catalog_request = if modules.is_empty() {
        serde_json::json!({})
    } else {
        serde_json::json!({ "roots": modules })
    };
    let catalog = invoke("catalog.methods", &catalog_request)?;
    if !catalog.outcome.carries_payload() {
        return Ok(catalog);
    }
    let methods = field(&catalog.payload, "methods")?;
    if methods.as_array().is_none_or(Vec::is_empty) {
        return Err("no active module declares a `verification_catalog`, so there is nothing to advise from".to_owned());
    }
    let store = invoke(
        "evidence.audit_inputs",
        &serde_json::json!({ "repo": repo }),
    )?;
    if !store.outcome.carries_payload() {
        return Ok(store);
    }
    let advised = invoke(
        "auditor.advise",
        &serde_json::json!({
            "catalog": catalog.payload,
            "obligations": field(&coverage.payload, "obligations")?,
            "shapes": field(&properties.payload, "shapes")?,
            "bindings": field(&store.payload, "bindings")?,
            "runs": field(&store.payload, "runs")?,
            "diagnostics": field(&coverage.payload, "diagnostics")?,
        }),
    )?;
    if !advised.outcome.carries_payload() {
        return Ok(advised);
    }
    let all = field(&advised.payload, "advice")?;
    let shown = all
        .as_array()
        .ok_or_else(|| "auditor.advise did not return advice".to_owned())?
        .iter()
        .filter(|advice| {
            (!arguments.get_flag("mismatch-only") && !arguments.get_flag("inconclusive-only"))
                || (arguments.get_flag("mismatch-only")
                    && advice
                        .get("mismatch")
                        .and_then(serde_json::Value::as_bool)
                        .unwrap_or(false))
                || (arguments.get_flag("inconclusive-only")
                    && advice
                        .get("inconclusive")
                        .and_then(serde_json::Value::as_bool)
                        .unwrap_or(false))
        })
        .cloned()
        .collect::<Vec<_>>();
    let rendered = if arguments.get_flag("json") {
        serde_json::to_string(&serde_json::json!({ "advice": shown }))
            .map_err(|error| error.to_string())?
    } else {
        render(&shown, all.as_array().map_or(&[], Vec::as_slice))?
    };
    Ok(Response::ok(serde_json::json!({ "rendered": rendered })))
}

fn field(value: &serde_json::Value, name: &str) -> Result<serde_json::Value, String> {
    value
        .get(name)
        .cloned()
        .ok_or_else(|| format!("core response did not return {name}"))
}
fn required(arguments: &ArgMatches, name: &str) -> Result<String, String> {
    arguments
        .get_one::<String>(name)
        .cloned()
        .ok_or_else(|| format!("--{name} is required"))
}
fn render(shown: &[serde_json::Value], all: &[serde_json::Value]) -> Result<String, String> {
    let mut lines = Vec::with_capacity(shown.len() + 2);
    for advice in shown {
        let inconclusive = advice
            .get("inconclusive")
            .and_then(serde_json::Value::as_bool)
            .unwrap_or(false);
        let conclusion = if inconclusive {
            "inconclusive".to_owned()
        } else {
            format!("recommend: {}", recommendations(advice)?)
        };
        lines.push(format!(
            "{}  authored={}  {}{}{}",
            advice
                .get("obligation")
                .and_then(serde_json::Value::as_str)
                .unwrap_or(""),
            advice
                .get("authored")
                .and_then(serde_json::Value::as_str)
                .unwrap_or("—"),
            conclusion,
            if advice
                .get("mismatch")
                .and_then(serde_json::Value::as_bool)
                .unwrap_or(false)
            {
                "  ⚠ mismatch"
            } else {
                ""
            },
            if advice
                .get("uncatalogued")
                .and_then(serde_json::Value::as_bool)
                .unwrap_or(false)
            {
                "  ⚠ uncatalogued"
            } else {
                ""
            }
        ));
    }
    let count = |field: &str| {
        all.iter()
            .filter(|advice| {
                advice
                    .get(field)
                    .and_then(serde_json::Value::as_bool)
                    .unwrap_or(false)
            })
            .count()
    };
    lines.push(String::new());
    lines.push(format!("{} of {} obligation(s) shown. Of all {}: {} mismatch, {} uncatalogued, {} inconclusive. Recommendations, not verdicts: confirm the method in spec review.", shown.len(), all.len(), all.len(), count("mismatch"), count("uncatalogued"), count("inconclusive")));
    Ok(lines.join("\n"))
}

/// Render the first three recommendation rows exactly as the retained shell.
fn recommendations(advice: &serde_json::Value) -> Result<String, String> {
    let rows = advice
        .get("recommended")
        .and_then(serde_json::Value::as_array)
        .ok_or_else(|| "auditor.advise entry did not return recommendations".to_owned())?;
    rows.iter()
        .take(3)
        .map(|row| {
            let method = row
                .get("method")
                .and_then(serde_json::Value::as_str)
                .ok_or_else(|| "advice recommendation did not return method".to_owned())?;
            let reasons = row
                .get("reasons")
                .and_then(serde_json::Value::as_array)
                .ok_or_else(|| "advice recommendation did not return reasons".to_owned())?
                .iter()
                .map(|reason| {
                    reason
                        .get("value")
                        .and_then(serde_json::Value::as_str)
                        .ok_or_else(|| {
                            "advice recommendation reason did not return value".to_owned()
                        })
                })
                .collect::<Result<Vec<_>, _>>()?;
            Ok(format!("{method} ({})", reasons.join(", ")))
        })
        .collect::<Result<Vec<_>, String>>()
        .map(|rows| rows.join("; "))
}

#[cfg(test)]
#[allow(clippy::expect_used, reason = "test fixtures may panic")]
mod tests {
    use super::render;

    /// Trace: FR-031, FR-062, FR-102
    #[test]
    fn tc_373_advice_human_output_keeps_methods_reasons_and_the_three_row_limit() {
        let rows = serde_json::json!([
            {
                "obligation": "FR-001-AC-1",
                "authored": "Test",
                "inconclusive": false,
                "mismatch": true,
                "uncatalogued": false,
                "recommended": [
                    { "method": "Analysis", "reasons": [{ "value": "characteristic" }, { "value": "property" }] },
                    { "method": "Test", "reasons": [{ "value": "risk" }] },
                    { "method": "Inspection", "reasons": [{ "value": "review" }] },
                    { "method": "Demonstration", "reasons": [{ "value": "ignored" }] }
                ]
            },
            {
                "obligation": "NFR-001",
                "authored": null,
                "inconclusive": true,
                "mismatch": false,
                "uncatalogued": true,
                "recommended": []
            }
        ]);
        let rows = rows.as_array().expect("fixture is an advice array");

        assert_eq!(
            render(rows, rows).expect("valid advice renders"),
            "FR-001-AC-1  authored=Test  recommend: Analysis (characteristic, property); Test (risk); Inspection (review)  ⚠ mismatch\nNFR-001  authored=—  inconclusive  ⚠ uncatalogued\n\n2 of 2 obligation(s) shown. Of all 2: 1 mismatch, 1 uncatalogued, 1 inconclusive. Recommendations, not verdicts: confirm the method in spec review."
        );
    }
}
